### PR TITLE
[IMT] Let TThreadedObject grow the number of slots as needed 

### PR DESCRIPTION
**Pull request by eguiraud**
TThreadedObject currently has a fixed amount of available slots,
the number of which can be set at construction time. This makes
it impossible to use TThreadedObject inside TBB tasks safely, because
TBB can potentially swap out worker threads for different threads at
runtime, resulting in TThreadedObject counting more thread ID's than
slots, and therefore in out-of-bound accesses to its contents.

With this patch, TThreadedObject just appends a new slot when needed,
so it nevers runs out.

### CODE REVIEW

File: core/thread/inc/ROOT/TThreadedObject.hxx [hunk 1]
```diff
@@ -38,22 +40,15 @@ namespace ROOT {
     * \brief Defines the number of threads in some of ROOT's interfaces.
     */
    struct TNumSlots {
-      int fVal; // number of slots
+      unsigned int fVal; // number of slots
```
Comment (Axel-Naumann) [inline core/thread/inc/ROOT/TThreadedObject.hxx:43]: `unsigned` arithmatics (`++`, `<=`) are a nightmare. Do we really need this?

File: core/thread/inc/ROOT/TThreadedObject.hxx [hunk 2]
```diff
@@ -100,34 +95,26 @@ namespace ROOT {
...
+         template <class T>
+         struct DirCreator<T, true> {
+            static TDirectory *Create() { return nullptr; }
          };
 
          struct MaxSlots_t {
             unsigned fVal;
             operator unsigned() const { return fVal; }
             MaxSlots_t &operator=(unsigned val)
-               R__DEPRECATED(6, 24, "Superior interface exists: construct TThreadedObject using TNumSlots instead.")
```
Comment (Axel-Naumann) [inline core/thread/inc/ROOT/TThreadedObject.hxx:130]: `R__DEPRECATED` is the user API :-)

File: core/thread/inc/ROOT/TThreadedObject.hxx [hunk 3]
```diff
@@ -173,84 +161,112 @@ namespace ROOT {
...
-            Warning("TThreadedObject::GetAtSlot", "Maximum number of slots reached.");
+         std::size_t nAvailableSlots;
+         {
+            // fObjPointers can grow due to a concurrent operation on this TThreadedObject, need to lock
+            std::lock_guard<ROOT::TSpinMutex> lg(fSpinMutex);
+            nAvailableSlots = fObjPointers.size();
+         }
+
+         if (i >= nAvailableSlots) {
+            Warning("TThreadedObject::GetAtSlot", "This slot does not exist.");
```
Comment (Axel-Naumann) [inline core/thread/inc/ROOT/TThreadedObject.hxx:222]: So then in the end this doesn't grow the slots? Should it not?

File: core/thread/inc/ROOT/TThreadedObject.hxx (hunk 1 above)
Comment (eguiraud) [reply core/thread/inc/ROOT/TThreadedObject.hxx:43]: > Do we really need this?

As much as you ever really need unsigned integers...it's nice that `TNumSlots{-1}` simply does not compile and I don't have to add a ctor that checks for non-negativity. Plus the number of slots must be non-negative, always grows, and should stay quite far from UINT_MAX. What kind of problems do you expect?

(just let me know if you want an `int`)

Comment (Axel-Naumann) [reply core/thread/inc/ROOT/TThreadedObject.hxx:43]: OK. But comparison etc is evil - the general (committee) recommendation is "don't use unsigned unless you do bit manipulations". https://wesmckinney.com/blog/avoid-unsigned-integers/ https://www.boost.org/doc/libs/develop/libs/safe_numerics/doc/html/tutorial/3.html https://github.com/ericniebler/stl2/issues/182 :-) Your call, I get your argument.

File: core/thread/inc/ROOT/TThreadedObject.hxx (hunk 3 above)
Comment (eguiraud) [reply core/thread/inc/ROOT/TThreadedObject.hxx:222]: Good point.

With the patch in its current form, `TThreadedObject` only grows when access is performed with `Get` or `->` (which make up the simple, sane interface).

When a user asks for a specific slot number through `GetAtSlot`, the problem is that when user requires slot `i`, I would have to create all slots from `nSlots` to `i` included. Or change TThreadedObject's internals a bit more to allow "holes". The other problem is that I am not sure what the expected kind of usage for `GetAtSlot` is.

These are the two usage patterns that I foresaw for `TThreadedObject`:

1. user creates `TThreadedObject(model)` and then has however many threads concurrently accessing its contents via the arrow operator. Each thread gets its own copy of `model`, and new slots are appended as needed (this solves our problem with `TTreeProcessorMT`)
2. as a performance optimization, user can create `TThreadedObject(TNumSlots{4}, model)` and then set the slot contents via `SetAtSlot` or modify the contents via `GetAtSlot`. This is, of course, if they are sure that they won't use more than `TNumSlots{4}` slots _or_ they let threads access slots concurrently via `Get` or `->`, which grows the slots as needed, but creates new slots copying the model, does not let users choose their initial content. We don't have this usage pattern in ROOT afaik

P.S.
I broke a use-case: user creates `TThreadedObject(model)` and then accesses `GetAtSlot(i)` with `i <= fgMaxSlots`: before, `GetAtSlot` would lazily populate the slot number requested with a copy of `model`, now we just return a `nullptr`. Fixing it now.

File: core/thread/inc/ROOT/TThreadedObject.hxx (hunk 2 above)
Comment (eguiraud) [reply core/thread/inc/ROOT/TThreadedObject.hxx:130]: will fix (here and elsewhere, also my fault :D)

File: core/thread/inc/ROOT/TThreadedObject.hxx (hunk 3 above)
Comment (eguiraud) [reply core/thread/inc/ROOT/TThreadedObject.hxx:222]: Forgot to ask questions:
* do you think `GetAtSlot(i)` should grow the number of (empty, nullptr) slots to "at least i"?
* how is `GetAtSlot` supposed to be used?

Comment (eguiraud) [reply core/thread/inc/ROOT/TThreadedObject.hxx:223]: Agreed with Axel that it's fine if `GetAtSlot(i)` throws if `i > currentNumberOfSlots`.

**Pull request merged**